x-pack/plugins/stack_connectors/server/connector_types/slack/index.ts

```diff
@@ -28,114 +20,79 @@ import {
 import { renderMustacheString } from '@kbn/actions-plugin/server/lib/mustache_renderer';
 import { getCustomAgents } from '@kbn/actions-plugin/server/lib/get_custom_agents';
 import { getRetryAfterIntervalFromHeaders } from '../lib/http_response_retry_header';
+import type {
+  SlackWebApiExecutorOptions,
+  SlackWebhookExecutorOptions,
+  WebhookParams,
+  SlackExecutorOptions,
+  SlackConnectorType,
+  WebApiParams,
+} from '../../../common/slack/types';
+import {
+  SlackSecretsSchema,
+  SlackParamsSchema,
+  SlackConfigSchema,
+} from '../../../common/slack/schema';
+import { SLACK_CONNECTOR_ID } from '../../../common/slack/constants';
+import { SLACK_CONNECTOR_NAME } from './translations';
+import { api } from './api';
+import { createExternalService } from './service';
+import { validate } from './validators';
 
-export type SlackConnectorType = ConnectorType<
-  {},
-  ConnectorTypeSecretsType,
-  ActionParamsType,
-  unknown
->;
-export type SlackConnectorTypeExecutorOptions = ConnectorTypeExecutorOptions<
-  {},
-  ConnectorTypeSecretsType,
-  ActionParamsType
->;
-
-// secrets definition
-
-export type ConnectorTypeSecretsType = TypeOf<typeof SecretsSchema>;
-
-const secretsSchemaProps = {
-  webhookUrl: schema.string(),
-};
-const SecretsSchema = schema.object(secretsSchemaProps);
-
-// params definition
-
-export type ActionParamsType = TypeOf<typeof ParamsSchema>;
-
-const ParamsSchema = schema.object({
-  message: schema.string({ minLength: 1 }),
-});
-
-// connector type definition
-
-export const ConnectorTypeId = '.slack';
-// customizing executor is only used for tests
-export function getConnectorType({
-  executor = slackExecutor,
-}: {
-  executor?: ExecutorType<{}, ConnectorTypeSecretsType, ActionParamsType, unknown>;
-}): SlackConnectorType {
+export function getConnectorType(): SlackConnectorType {
   return {
-    id: ConnectorTypeId,
+    id: SLACK_CONNECTOR_ID,
     minimumLicenseRequired: 'gold',
-    name: i18n.translate('xpack.stackConnectors.slack.title', {
-      defaultMessage: 'Slack',
-    }),
+    name: SLACK_CONNECTOR_NAME,
     supportedFeatureIds: [
       AlertingConnectorFeatureId,
       UptimeConnectorFeatureId,
       SecurityConnectorFeatureId,
     ],
     validate: {
+      config: {
+        schema: SlackConfigSchema,
+      },
       secrets: {
-        schema: SecretsSchema,
-        customValidator: validateConnectorTypeConfig,
+        schema: SlackSecretsSchema,
+        customValidator: validate.secrets,
       },
       params: {
-        schema: ParamsSchema,
+        schema: SlackParamsSchema,
       },
+      connector: validate.connector,
     },
     renderParameterTemplates,
-    executor,
+    executor: (execOptions: SlackExecutorOptions) => {
+      const res =
+        execOptions.config.type === 'webhook'
+          ? slackWebhookExecutor(execOptions as SlackWebhookExecutorOptions)
```

Comment:
I think we still have a problem because the params could be for web_api but the config/secrets for webhook, or vice versa.  Since I don't see us validating the config/secrets/params all together.

We should do it here, or in the `slackXyzExecutor()` functions.  I'd suggest adding a new validation function in the `validators.ts` that validates the parameters based on the `config.type`, or it could be more generic and take the config, secrets and params.  Actually, maybe we can add an optional param to the existing `validateConnector()` call, for the params, and do the params validation as well IF it's passed in.  Making it an optional parameter seems like it would still allow us to use it as the `validate.connector()` in the connector type, but call it with the param during execution.

Reply:
@pmuellr Look, please. Does it looks like what'd you mean: 6596bf619a46a0368f34b23ef91c59c6570e669e?